# Roads Table Audit

## Objective

Perform a data quality assessment of the Roads dataset before loading it into the Silver Layer.

The audit focuses on:

- Missing values
- Duplicate records
- Primary key validation
- Data type validation
- Category standardization
- Business rule validation

In [1]:
import pandas as pd

## 1. Load Dataset

Load the Roads dataset into a Pandas DataFrame and perform an initial inspection.

In [2]:
roads = pd.read_csv(r"C:\Dataset\roads.csv")

## 2. Dataset Structure Review

Review:

- Number of records
- Number of columns
- Data types
- Overall dataset structure

In [3]:
roads.info

<bound method DataFrame.info of       road_id    road_type  num_lanes  num_signals    zone_type
0     RD00001       urban         2.0          8.0  residential
1     RD00002  residential        1.0          NaN  RESIDENTIAL
2     RD00003  residential        1.0          8.0  Residential
3     RD00004       urban         3.0          9.0  Residential
4     RD00005      HIGHWAY        4.0          3.0   commerical
...       ...          ...        ...          ...          ...
9995  RD09996  residential        2.0          9.0   commerical
9996  RD09997      resdntl        1.0         15.0   Commercial
9997  RD09998  Residential        1.0          5.0   commerical
9998  RD09999        urban        3.0          6.0   commercial
9999  RD10000       urban         2.0          5.0   commerical

[10000 rows x 5 columns]>

In [4]:
roads.dtypes

road_id         object
road_type       object
num_lanes      float64
num_signals    float64
zone_type       object
dtype: object

## 3. Missing Value Analysis

Identify missing values across all columns and measure their impact on data completeness.

In [5]:
roads.isnull().sum()

road_id        300
road_type        0
num_lanes      700
num_signals    600
zone_type      800
dtype: int64

## 4. Duplicate Record Analysis

Check for fully duplicated rows that may have resulted from ingestion or processing issues.

In [6]:
roads.duplicated().sum()

np.int64(8)

## 5. Primary Key Validation

Evaluate whether `road_id` can serve as a reliable primary key by validating:

- NULL values
- Uniqueness
- Duplicate occurrences

In [7]:
roads['road_id'].nunique()

9341

In [8]:
roads['road_id'].duplicated().sum()

np.int64(658)

## 6. Road Type Standardization Review

Inspect categorical values for:

- Inconsistent casing
- Spelling variations
- Invalid categories

In [9]:
roads['road_type'].unique()

array(['urban ', 'residential', 'HIGHWAY', 'Highway', 'Residential',
       'Hghway', 'Resdntl', 'urbn', 'highway', 'urban', 'Urban', 'Urbn',
       'residental', 'hghway', 'resdntl', 'URBAN', 'RESIDENTIAL'],
      dtype=object)

In [10]:
roads['road_type'].nunique()

17

## 7. Numeric Data Validation

Validate numeric columns against expected business rules and identify invalid values.

In [11]:
roads[roads['num_lanes'] < 0].count()

road_id        60
road_type      61
num_lanes      61
num_signals    58
zone_type      56
dtype: int64

In [12]:
roads[roads['num_signals'] < 0].count()

road_id        0
road_type      0
num_lanes      0
num_signals    0
zone_type      0
dtype: int64

## 8. Zone Type Standardization Review

Review zone categories for:

- Missing values
- Inconsistent casing
- Spelling variations

In [13]:
roads['zone_type'].unique()

array(['residential', 'RESIDENTIAL', 'Residential', 'commerical',
       'industrial', 'Commercial', nan, 'COMMERCIAL', 'INDUSTRIAL',
       'industral', 'commercial', 'Industrial', 'residental'],
      dtype=object)

In [14]:
roads['zone_type'].nunique()

12

In [15]:
roads[roads['road_id'].duplicated()].count()

road_id        359
road_type      658
num_lanes      614
num_signals    620
zone_type      606
dtype: int64

## 9. Audit Findings Summary

### Key Findings

- 300 NULL road_id values
- 658 duplicated road_id values
- 700 missing num_lanes values
- 61 negative num_lanes values
- Inconsistent road_type categories
- Inconsistent zone_type categories

### Candidate Primary Key Status

❌ road_id cannot currently be used as a primary key.

### Recommended Silver Layer Actions

- Remove duplicate rows
- Resolve duplicate road_id values
- Handle missing values
- Standardize categorical values
- Validate business categories